packages/lesswrong/server/useractivities/cron.ts

```diff
@@ -0,0 +1,293 @@
+/* eslint-disable no-console */
+import { chunk, max } from 'lodash/fp';
+import UserActivities from '../../lib/collections/useractivities/collection';
+import { randomId } from '../../lib/random';
+import { getSqlClientOrThrow } from '../../lib/sql/sqlClient';
+import { addCronJob } from '../cronUtil';
+import { Vulcan } from '../vulcan-lib';
+import { ActivityWindowData, getUserActivityData } from './getUserActivityData';
+
+const ACTIVITY_WINDOW_HOURS = 21 * 24; // 3 weeks
+
+/**
+ * Assert that all the activityArrays in the UserActivities table are the correct (and identical) length,
+ * and have the correct start and end dates. Drop any rows that are inconsistent.
+ *
+ * In practice this should never do anything, but it will stop the data getting badly out of sync if
+ * something goes wrong.
+ */
+async function assertTableIntegrity(dataDb: SqlClient) {
+  // Step 1: Check if all rows have the same startDate and endDate
+  const dateCheckResult = await dataDb.one(`
+    SELECT COUNT(DISTINCT "startDate") AS start_count, COUNT(DISTINCT "endDate") AS end_count
+    FROM "UserActivities";
+  `);
+
+  if (dateCheckResult.start_count > 1 || dateCheckResult.end_count > 1) {
+    // eslint-disable-next-line no-console
+    console.error('UserActivities table has rows with different start and end dates. Dropping rows to fix this.');
+
+    // Delete rows with different startDate and endDate
+    await dataDb.none(`
+      DELETE FROM "UserActivities"
+      WHERE "startDate" <> (SELECT MIN("startDate") FROM "UserActivities")
+         OR "endDate" <> (SELECT MIN("endDate") FROM "UserActivities");
+    `);
+  }
+
+  // Step 2: Check if the array of activity has the correct length for each row
```

Comment:
It shouldn't ever happen in practice, but in principle they could be set differently due to a bug

Reply:
I see the value of this whole function as mainly a developer aid rather than actually expecting these DELETEs to do anything in prod. It's just that if you keep doing complex appends and concatenations over multiple days without ever asserting on the entire state it will become hard to trust the data